cirq-core/cirq/ops/matrix_mixture.py

```diff
@@ -0,0 +1,106 @@
+from typing import Any, Dict, Iterable, Optional, Tuple
+import numpy as np
+
+from cirq import protocols, value
+from cirq._compat import proper_repr
+from cirq.ops import raw_types
+
+
+class MatrixMixture(raw_types.Gate):
+    """A generic mixture that can record the index of its selected operator.
+
+    Args:
+        mixture: a list of (probability, unitary) pairs
+        key: an optional measurement key string for this mixture. Simulations
+            which select a single unitary to apply will store the index
+            of that unitary in the measurement result list with this key.
+    """
+
+    def __init__(self, mixture: Iterable[Tuple[float, np.ndarray]], key: Optional[str] = None):
+        mixture = list(mixture)
+        if not mixture:
+            raise ValueError('MatrixMixture must have at least one unitary.')
+        if not protocols.approx_eq(sum(p[0] for p in mixture), 1):
+            raise ValueError('Unitary probabilities must sum to 1.')
+        m0 = mixture[0][1]
+        num_qubits = np.log2(m0.size) / 2
+        if not num_qubits.is_integer():
+            raise ValueError(
+                f'Input mixture of shape {m0.shape} does not represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
+        for i, op in enumerate(p[1] for p in mixture):
+            if not op.size == m0.size:
+                raise ValueError(
+                    f'Inconsistent unitary sizes: op[0]: {m0.size}, op[{i}]: {op.size}'
+                )
```

Comment:
Do we want to verify the matrix is unitary (similar to above) ?

Reply:
Done.